sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -151,6 +164,30 @@ abstract class Expression extends TreeNode[Expression] {
   }
 
   /**
+   * Returns the hash for this expression. Expressions that compute the same result, even if
+   * they differ cosmetically should return the same hash.
+   */
+  def semanticHash() : Int = {
```

Comment:
Sorry, I am still confusing, why we can't use the `hashCode` instead.
The motivation of `semanticEquals` is to ignore the `AttributeReference.name` in comparison for `AttributeReference`, and the `AttributeReference.hashCode` also does ignore the `AttributeReference.name`, I don't think the `cosmetic differences` really exists for `hashCode`. Isn't it?